### DISCUSSION TITLE
Diagnostics popup rendered as Markdown callouts

### DISCUSSION THREAD

**Discussion by Picalines** [MeanderingProgrammer/render-markdown.nvim#224] [Show and tell]
I had fun rendering neovim diagnostics as md callouts with `render-markdown`

<img width="585" alt="image" src="https://github.com/user-attachments/assets/0183eaef-4ef8-4b0f-99fa-8dc1462dde15">

It's useful if you transform your LSP messages to MD. I've combined it with [format-ts-errors](https://github.com/davidosomething/format-ts-errors.nvim/issues/6):

<img width="492" alt="image" src="https://github.com/user-attachments/assets/7f7f8e17-fe03-42eb-a9e0-d799144130c4">

It might be too much for some people, but here's the code if you want:

```lua
local severity_to_md_quote = {
	[vim.diagnostic.severity.ERROR] = '> [!ERROR]',
	[vim.diagnostic.severity.WARN] = '> [!WARNING]',
	[vim.diagnostic.severity.INFO] = '> [!INFO]',
	[vim.diagnostic.severity.HINT] = '> [!HINT]',
}

vim.diagnostic.config {
	float = {
		-- NOTE: i had some weird results using `format`, prefix & suffix work better
		format = function()
			return ''
		end,

		prefix = function(report)
			return severity_to_md_quote[report.severity] .. string.format(' %s\n', report.code), 'Normal'
		end,

		suffix = function(report, i, total)
			local message = vim.trim(report.user_data.lsp.message)
			local md_quote = '> ' .. table.concat(vim.split(message, '\n'), '\n> ')
			local separator = '\n' .. (i < total and '---' or '')
			return md_quote .. separator, 'Normal'
		end,
	},
}

local builtin_open_float = vim.diagnostic.open_float

---@diagnostic disable-next-line: duplicate-set-field
vim.diagnostic.open_float = function(...)
	local bufnr, winid = builtin_open_float(...)

	if bufnr and winid then
		vim.api.nvim_buf_clear_namespace(bufnr, -1, 0, -1) -- clears builtin extmarks
		vim.api.nvim_set_option_value('filetype', 'markdown', { buf = bufnr })
	end

	return bufnr, winid
end
```

**Comment by MeanderingProgrammer** [comment]
Thanks for the post, this is a neat use case :)

I looked at your issue in the `format-ts-errors.nvim` repo and was a little confused when you mentioned that this plugin `removes the indentation in code blocks`. This plugin will add a background to code blocks but it shouldn't remove any indentation.
